### PR TITLE
Tacking clipper

### PR DESCRIPTION
**Pull request by kmbarry1**
Still considering the best way to run this through all the tests for Clippers in `dss` without copying that entire test file to this repo...factoring those into a "Clipper conformance tests" suite somewhere might be the right way to go.

But you can diff the last commit that changed the tacking Clipper against the initial commit (which was a straight copy-paste) to easily see just the changes I made.

### CODE REVIEW

**Comment by kmbarry1** [comment]
Use this link to compare `cropper.sol` changes against the original Clipper:
https://github.com/makerdao/dss-crop-join/compare/910e0374e5bb2e35cfb0267180229bc579fe6bb7...78d89f7817da3ac16661edd8e902896e4b240606

**Review by hexonaut — CHANGES_REQUESTED** [review]
When reviewing this I have come across a pretty critical assumption that is pretty easy to violate. We are assuming `gems <= stake[usr]` at all times, but this can be pretty easily circumvented by having a user `flux` their gems without calling `tack`. I've added this problematic case as a failing test in the subsequent commit. Even if we remove the `tack()` calls from the liquidations we still have this problem where people could move gems around in a quite complex pattern and we leave it up to the keepers to tack these around from arbitrary addresses.

I think we need to do 2 things to solve this issue.

1. Liquidations should not do anything fancy with the rewards. I think upon `kick` being called any outstanding rewards should just be sent to the vault owner (or maybe we just do nothing and the rewards are lost).
2. `flee()` and `exit()` need to allow for the case where `tack()` was not called and you just get the collateral back without rewards.

File: src/cropper.sol [hunk 1]
```diff
@@ -0,0 +1,478 @@
...
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity 0.6.12;
+
+import "./crop.sol";
```
[the remark this answers is not in the document]
Comment (kmbarry1) [reply src/cropper.sol:19]: > When reviewing this I have come across a pretty critical assumption that is pretty easy to violate. We are assuming `gems <= stake[usr]` at all times, but this can be pretty easily circumvented by having a user `flux` their gems without calling `tack`. I've added this problematic case as a failing test in the subsequent commit. Even if we remove the `tack()` calls from the liquidations we still have this problem where people could move gems around in a quite complex pattern and we leave it up to the keepers to tack these around from arbitrary addresses.
> 
> I think we need to do 2 things to solve this issue.
> 
> 1. Liquidations should not do anything fancy with the rewards. I think upon `kick` being called any outstanding rewards should just be sent to the vault owner (or maybe we just do nothing and the rewards are lost).
> 2. `flee()` and `exit()` need to allow for the case where `tack()` was not called and you just get the collateral back without rewards.

Yeah good catch. My thoughts:

1. Sending any rewards to the Vault owner on `kick` is fair--the rewards don't collateralize their position, so there's no real reason someone should lose those rewards if liquidated. The difficulty is the CDPManager thing Gonzalo brought up--we can really only send rewards back to the urn address, i.e. the UrnHandler, and the rewards would be lost anyway. The only fully satisfactory solution I see is unfortunately quite complex: the adapter itself should be something like the CdpManager, provisioning internally some address to control the Vault, and only exposing a subset of Vat UIs to end users (so e.g. external users could only `flux` via a function on the adapter that would enforce appropriate `tack` usage).

2. For `exit` and `flee`...need to think carefully here that there isn't e.g. some griefing vector. But if we went with the design above (adapter becomes a middleware that mediates access to the core APIs), the problem goes away here too.

Basically I now think the way we should have done it in the first place is what I describe in (1), but we're pretty far down this other path so there's an argument we should accept the suboptimal thing to get this out the door. I might spin off another branch to try out that idea and see just how complex it is.
